Assets/Scripts/MineManager.cs

```diff
@@ -0,0 +1,221 @@
+                {
+                    gameGrid[i, j] = CheckNeighbor(new Vector2Int(i, j));
+                }
+            }
+        }
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        if (endGame)
+        {
+            time += Time.deltaTime;
+            if (time > delayTime)
+            {
+                endGame = false;
+                time = 0f;
+                SceneManager.LoadScene("Scenes/End");
+            }
+        }
+    }
+    
+    int CheckNeighbor(Vector2Int loc)
+    {
+        int counter = 0;
+        for (int i = -1; i < 2; i++)
+        {
+            if (loc.x + i < 0 || loc.x + i >= gridSize)
+            {
+                continue;
+            }
+            for (int j = -1; j < 2; j++)
+            {
+                if (loc.y + j >= 0 && loc.y + j < gridSize)
+                {
+                    if (gameGrid[loc.x + i, loc.y + j] ==-1)
+                    {
+                        counter += 1;
+                    }
+                }
+            }
+        }
+        return counter;
+    }
+
+    public void ChangeState(Vector2Int loc, State state)
+    {
+        gridStates[loc.x, loc.y] = state;
+        if (state == State.Opened)
+        {
+            opened += 1;
+            Debug.Log(loc+": "+opened);
+            if (gameGrid[loc.x, loc.y] == -1)
+            {
+                Debug.Log("end game lose");
+                endGame = true;
+                lose = true;
+            }
+            if (opened == gridSize * gridSize - mineCount)
+            {
+                Debug.Log("end game win");
+                endGame = true;
+                win = true;
+            }
+        }
+    }
+    
+    public void ChangeSpriteFlag(Vector2Int loc, GameObject o)
+    {
+        o.GetComponent<SpriteRenderer>().sprite = flag;
+        o.GetComponent<SpriteRenderer>().color = Color.white;
+        o.transform.localScale = new Vector3((1024f / (gridSize+1)) / flag.rect.size.x,
+            (1024f / (gridSize+1)) / flag.rect.size.y, 1f);
+    }
+    
+                
+    public void ChangeSprite(Vector2Int loc, GameObject o)
+    {
+        if (gameGrid[loc.x, loc.y] == -1)
+        {
+            o.GetComponent<SpriteRenderer>().sprite = flag;
+            o.GetComponent<SpriteRenderer>().color = Color.red;
+            o.transform.localScale = new Vector3((1024f / (gridSize+1)) / flag.rect.size.x,
+                (1024f / (gridSize+1)) / flag.rect.size.y, 1f);
+        }
+        else
+        {
+            Sprite image = number[gameGrid[loc.x,loc.y]];
+            o.GetComponent<SpriteRenderer>().sprite = image;
+            o.GetComponent<SpriteRenderer>().color = Color.white;
+            o.transform.localScale = new Vector3((1024f / (gridSize+1)) / image.rect.size.x,
+                (1024f / (gridSize+1)) / image.rect.size.y, 1f);
+        }
+    }
+    public void ChangeSpriteReset(Vector2Int loc, GameObject o)
```

Comment:
DRY. there should just be a sinlge function that updates based on the data from model. you arent doing that.